java/ql/src/semmle/code/java/security/CommandArguments.qll

```diff
@@ -0,0 +1,217 @@
+
+/**
+ * Holds if `ex` is used safely as an argument to a command;
+ * i.e. it's not in the first position and it's not a shell command.
+ */
+predicate isSafeCommandArgument(Expr ex) {
+  exists(ArrayInit ai, int i |
+    ex = ai.getInit(i) and
+    i > 0 and
+    not isShell(ai.getInit(0))
+  )
+  or
+  exists(CommandArgumentList cal |
+    cal.isNotShell() and
+    ex = cal.getASubsequentAdd().getArgument(0)
+  )
+  or
+  exists(CommandArgumentArray caa |
+    caa.isNotShell() and
+    ex = caa.getAWrite(any(int i | i > 0))
+  )
+}
+
+/**
+ * Holds if the given expression is the name of a shell command such as bash or python
+ */
+private predicate isShell(Expr ex) {
+  exists(string cmd | cmd = ex.(StringLiteral).getValue() |
+    cmd.regexpMatch(".*(sh|javac?|python[23]?|osascript|cmd)(\\.exe)?$")
+  )
+  or
+  exists(SsaVariable ssa |
+    ex = ssa.getAUse() and
+    isShell(ssa.getAnUltimateDefinition().(SsaExplicitUpdate).getDefiningExpr())
+  )
+  or
+  isShell(ex.(Assignment).getRhs())
+  or
+  isShell(ex.(LocalVariableDeclExpr).getInit())
+}
+
+/**
+ * A type that could be a list of strings. Includes raw `List` types.
+ */
+private class ListOfStringType extends CollectionType {
+  ListOfStringType() {
+    this.getSourceDeclaration().getASourceSupertype*().hasQualifiedName("java.util", "List") and
+    this.getElementType().getASubtype*() instanceof TypeString
+  }
+}
+
+/**
+ * A variable that could be used as a list of arguments to a command.
+ */
+private class CommandArgumentList extends SsaExplicitUpdate {
+  CommandArgumentList() {
+    this.getSourceVariable().getType() instanceof ListOfStringType and
+    forex(CollectionMutation ma | ma.getQualifier() = this.getAUse() |
+      ma.getMethod().getName().matches("add%")
+    )
+  }
+
+  /** Gets a use of the veriable for which the list could be empty. */
+  private RValue getAUseBeforeFirstAdd() {
+    result = getAFirstUse()
+    or
+    exists(RValue mid |
+      mid = getAUseBeforeFirstAdd() and
+      adjacentUseUse(mid, result) and
+      not exists(MethodAccess ma |
+        mid = ma.getQualifier() and
+        ma.getMethod().hasName("add")
+      )
+    )
+  }
+
+  /**
+   * Gets an addition to this list, i.e. a call to an `add` or `addAll` method.
+   */
+  MethodAccess getAnAdd() {
+    result.getQualifier() = getAUse() and
+    result.getMethod().getName().matches("add%")
+  }
+
+  /** Gets an addition to this list which could be its first element. */
+  MethodAccess getAFirstAdd() {
+    result = getAnAdd() and
+    result.getQualifier() = getAUseBeforeFirstAdd()
+  }
+
+  /** Gets an addition to this list which is not the first element. */
+  MethodAccess getASubsequentAdd() {
+    result = getAnAdd() and
+    not result = getAFirstAdd()
+  }
+
+  /** Holds if the first element of this list isn't a shell command. */
+  predicate isNotShell() {
+    forex(MethodAccess ma | ma = getAFirstAdd() | not isShell(ma.getArgument(0)))
+  }
```

Comment:
There are a lot no negations here. I think this would be simpler as:
```
predicate isShell() {
  exists(MethodAccess ma | ma = getAFirstAdd() and isShell(ma.gerArgument(0)))
}
```
The negation is probably also a smaller predicate, which also points towards using `isShell` rather than `isNotShell` as the interface.

Reply:
Good point, this interface can be simpler.